File: sample/cpp/10_plasma_effect.cpp
#include "runtime/cpp/native/core/py_runtime.h"
#include "runtime/cpp/native/core/process_runtime.h"
#include "runtime/cpp/native/core/scope_exit.h"

#include "generated/built_in/io_ops.h"
#include "generated/std/math.h"
#include "generated/std/time.h"
#include "generated/utils/gif.h"

// 10: Sample that outputs a plasma effect as a GIF.

void run_10_plasma_effect() {
    int64 w = 320;
    int64 h = 240;
    int64 frames_n = 216;
    str out_path = "sample/out/10_plasma_effect.gif";
    
    float64 start = pytra::std::time::perf_counter();
    rc<list<bytes>> frames = rc_list_from_value(list<bytes>{});
    
    for (int64 t = 0; t < frames_n; ++t) {
        bytearray frame = bytearray(w * h);
        for (int64 y = 0; y < h; ++y) {
            int64 row_base = y * w;
            for (int64 x = 0; x < w; ++x) {
                int64 dx = x - 160;
                int64 dy = y - 120;
                float64 v = pytra::std::math::sin((static_cast<float64>(x) + static_cast<float64>(t) * 2.0) * 0.045) + pytra::std::math::sin((static_cast<float64>(y) - static_cast<float64>(t) * 1.2) * 0.05) + pytra::std::math::sin((static_cast<float64>(x + y) + static_cast<float64>(t) * 1.7) * 0.03) + pytra::std::math::sin(pytra::std::math::sqrt(dx * dx + dy * dy) * 0.07 - static_cast<float64>(t) * 0.18);
                int64 c = static_cast<int64>((v + 4.0) * (255.0 / 8.0));
                if (c < 0)
                    c = 0;
                if (c > 255)
                    c = 255;
                frame[row_base + x] = c;
            }
        }
        rc_list_ref(frames).append(frame);
    }
    pytra::utils::gif::save_gif(out_path, w, h, rc_list_ref(frames), pytra::utils::gif::grayscale_palette(), 3, 0);
    float64 elapsed = pytra::std::time::perf_counter() - start;
    py_print("output:", out_path);
    py_print("frames:", frames_n);
    py_print("elapsed_sec:", elapsed);
}

int main(int argc, char** argv) {
    pytra_configure_from_argv(argc, argv);
    run_10_plasma_effect();
    return 0;
}

File: sample/py/10_plasma_effect.py
# 10: プラズマエフェクトをGIF出力するサンプル。

from __future__ import annotations

import math
from time import perf_counter

from pylib.gif_helper import grayscale_palette, save_gif


def run_10_plasma_effect() -> None:
    w = 320
    h = 240
    frames_n = 216
    out_path = "sample/out/10_plasma_effect.gif"

    start = perf_counter()
    frames: list[bytes] = []

    for t in range(frames_n):
        frame = bytearray(w * h)
        i = 0
        for y in range(h):
            for x in range(w):
                dx = x - 160
                dy = y - 120
                v = (
                    math.sin((x + t * 2.0) * 0.045)
                    + math.sin((y - t * 1.2) * 0.05)
                    + math.sin((x + y + t * 1.7) * 0.03)
                    + math.sin(math.sqrt(dx * dx + dy * dy) * 0.07 - t * 0.18)
                )
                c = int((v + 4.0) * (255.0 / 8.0))
                if c < 0:
                    c = 0
                if c > 255:
                    c = 255
                frame[i] = c
                i += 1
        frames.append(bytes(frame))

    save_gif(out_path, w, h, frames, grayscale_palette(), delay_cs=3, loop=0)
    elapsed = perf_counter() - start
    print("output:", out_path)
    print("frames:", frames_n)
    print("elapsed_sec:", elapsed)


if __name__ == "__main__":
    run_10_plasma_effect()
